src/newgrf.cpp

```diff
@@ -6211,6 +6397,56 @@ static void RoadStopMapSpriteGroup(ByteReader &buf, uint8_t idcount)
 	}
 }
 
+static void BadgeMapSpriteGroup(ByteReader &buf, uint8_t idcount)
+{
+	if (_cur.grffile->badge_map.empty()) {
+		GrfMsg(1, "BadgeMapSpriteGroup: No badges defined, skipping");
+		return;
+	}
+
+	std::vector<uint16_t> local_ids;
+	local_ids.reserve(idcount);
+	for (uint i = 0; i < idcount; i++) {
+		local_ids.push_back(buf.ReadExtendedByte());
+	}
+
+	uint8_t cidcount = buf.ReadByte();
+	for (uint c = 0; c < cidcount; c++) {
+		uint8_t ctype = buf.ReadByte();
+		uint16_t groupid = buf.ReadWord();
+		if (!IsValidGroupID(groupid, "BadgeMapSpriteGroup")) continue;
+
+		if (ctype >= GSF_END) continue;
+
+		for (const auto &local_id : local_ids) {
+			auto found = _cur.grffile->badge_map.find(local_id);
+			if (found == std::end(_cur.grffile->badge_map)) {
+				GrfMsg(1, "BadgeMapSpriteGroup: Badge {} undefined, skipping", local_id);
+				continue;
+			}
+
+			auto &badge = *GetBadge(found->second);
+			badge.grf_prop.SetSpriteGroup(ctype, _cur.spritegroups[groupid]);
+		}
+	}
```

Comment:
What does the `c` and c in `cidcount` stand for? Smells like 'cargo', but it's compared against `GrfSpecFeature`.

Reply:
"Yes", it's common code pattern used but duplicated by all features. `cid` and `cidcount` are the names in the Action 3 spec, although of course they derive from cargo types which was the usage originally.

These names are still used for (most) other features around here.